### DISCUSSION TITLE
Text File Check

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#198] [Q&A] [answered]
I'm trying to read in a text file (.json) and I want to do a simple check that the 1st 11 bytes match a standard pattern before importing it, as a level of security for not crashing everything in a screaming heap.

This is the code I tried, but it doesn't work (the :read(11)) is from io.read, which doesn't seem to be supported in ctrlr - it was a long shot).  Anyone know how I can do this check, or do I need to loadFileAsString to a variable and then check the variable?

```
if fileToRead:existsAsFile() then

		if fileToRead:read(11) ~= "\[{\"sysex\":\"" then
				utils.warnWindow(
					"Wrong File Type", 
					"This doesn't seem to be a Jupiter 8 Morph File, or it may have been corrupted. Loading this file will break stuff... badly, so we are aborting. Sorry!"
				)
			return
		end

```

**Comment by dobo365** [comment, marked as answer]
Are you trying to read a sysex or a text file?
If sysex then use this code:
```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
Load_OnChange = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- No action while loading a program or if the panel is in bootstrap or program states
	if bLoadingProgram or not isPanelReady() then
		return
	end	

--	Load_OnClick(mod:getComponent())

	-- Load a .syx file
	file = utils.openFileWindow ("Load program", File(""), "*.syx", true)
	if file:isValid() then	-- A file has been selected and the user pressed Open
		if file:existsAsFile() then
			if file:getSize()==600 then
				ProgramFileData = MemoryBlock()
				file:loadFileAsData(ProgramFileData)

				-- Check sysex header
				if ProgramFileData:getRange(0,8):toHexString(1) == "f0 00 20 32 00 01 aa 01" then

					-- do something...
				else
					utils.warnWindow ("Load program", "Error: not the expected program file")
				end
			else
				utils.warnWindow ("Load program", "Error: invalid file")
			end
		end
	end

end
```

**Reply by Godlike-Productions** [reply]
Thanks, it's a text file, but I'll try the same method.

**Comment by dnaldoog** [comment]
Are you worried whether the string "**sysex**" is in the message or just using that to make sure the file is from your program?

**Reply by Godlike-Productions** [reply]
No.

The import of json method will import and assign to my main data structure, regardless of the json.

I'm trying to do a check to ensure that what I'm importing is actually data that the plugin expects.

Those are the only characters that are constant in each export file

**Reply by dnaldoog** [reply]
So **sysex** will always start at position 4 in the string.

A safe way to check for **sysex** at that position is to use __string.find()__, then there's no need for all that funky escaping of the __[__ character.

<pre>
<code>
local str=fileToRead:loadFileAsString()
		if str:find('sysex') == 4 then
                    jsonDecode(str)
		else
		    utils.warnWindow("File Error","Invalid File")
		return
		end 
</pre>
</code>

**Reply by Godlike-Productions** [reply]
Yeah, I think I'll do that.  I was trying to do it before loading the file.  Trying to do a direct read, like the io library can.

I'll do it your way, once I get the export working.  Just about tracked down the nasty little bug.

Basically my json files were not populating because somehow all of the strings were turned into userdata, but they displayed and acted like strings.  I'm wondering if they were just passing on references.  Anyway, I'm up to fixing my 4th method now. Hopefully this is the end of it.

**Reply by dnaldoog** [reply]
Try  wrapping those string with L() <code>L(myString)</code>  function. Might fix it, but I don't really know the details - I think it fixes string incompatibility with lua.

**Reply by Godlike-Productions** [reply]
That was exactly the issue.  Unfortunately it was spread through a bunch of methods.  Now fixed 8.  I think that's all of them.  Fingers crossed.  Just testing now.

**Reply by Godlike-Productions** [reply]
Woohoo. It Lives.... It Lives....  Posting my load method now into the other thread.
